Receive and play a voice message

julioromano opened the pull request:
<!-- Please read [CONTRIBUTING.md](https://github.com/vector-im/element-x-android/blob/develop/CONTRIBUTING.md) before submitting your pull request -->
 
## Type of change

- [x] Feature
- [ ] Bugfix
- [ ] Technical
- [ ] Other :

## Content

This PR consists of several macro-blocks separated by path/package:
- `messages.impl.mediaplayer` : Global (room-wide) media player, now used only for voice messages but could be used for all media within EX in the future. It is backed by media3's exoplayer. Currently not unit-tested because mocking exoplayer is not trivial.
- `messages.impl.voicemessages.play` : Business logic of a timeline voice message. This is all the logic that manages the voice message bubble.
- `messages.impl.timeline.model` & `messages.impl.timeline.factories`: Timeline code that takes care of creating the `content` object for voice messages.
-  `messages.impl.timeline.components` : The actual View composable that shows the UI inside a voice message bubble.

All the rest is just small related changes that must be done here and there in existing code.

From a high level perspective this is how it works:
- Voice messages are unlike other message bubbles because they carry state (i.e. playing, downloading...) so they have a Presenter managing this state.
- Media content (i.e. the ogg file) of a voice message is downloaded from the rust SDK on first play then stored in a voice messages cache (see the `VoiceMessageCache` class, it is just a subdirectory in the app's cacheDir which is indexed by the matrix content uri). All further play attempts are done from the cache without hitting the rust SDK anymore.
- Playback of the ogg file is handled with the `VoiceMessagePlayer` class which is basically a "view" of the global `MediaPlayer` that allow the voice message to only see the media player state belonging to its media content. 
- Drawing of the waveform is done with an OSS library wrapped in the `WaveformProgressIndicator` composable.

Known issues:
 - The waveform has no position slider.
 - The waveform (and together with it the whole message bubble) is taller than the actual Figma design.
 - Swipe to reply for voice messages is disabled to avoid conflict with the audio scrubbing gesture (to reply to a voice message you have to use the long press menu).
 - The loading indicator is always shown (there is no delay).
 - Voice messages don't stop playing when redacted.

## Motivation and context

https://github.com/vector-im/element-meta/issues/2083

## Screenshots / GIFs

Provided by Screenshot tests in the PR itself.

DoM1niC commented:
earpiece listening would be nice to have, but works fine here on my fork!

julioromano commented:
> earpiece listening would be nice to have, but works fine here on my fork!

Thanks for the feedback!!!

julioromano commented on features/messages/impl/src/main/kotlin/io/element/android/features/messages/impl/timeline/TimelineItemPresenterFactories.kt:
```diff
@@ -0,0 +1,68 @@
+@Composable
+inline fun <reified C : TimelineItemEventContent, reified S : Any> TimelineItemPresenterFactories.rememberPresenter(
+    content: C
+): Presenter<S> = remember {
```
I suppose you mean keying by `content`?
Right now every voice message gets its own presenter so I'd think this is working as intended.
Is there anything else I should check to ensure we're fine without a key?

ganfra replied:
It's more about handling content updates

julioromano replied:
Done!

jonnyandrew commented on features/messages/impl/src/main/kotlin/io/element/android/features/messages/impl/voicemessages/timeline/VoiceMessagePresenter.kt:
```diff
@@ -0,0 +1,152 @@
+        fun eventSink(event: VoiceMessageEvents) {
+            when (event) {
+                is VoiceMessageEvents.PlayPause -> {
+                    if (playerState.isMyMedia) {
+                        if (playerState.isPlaying) {
+                            player.pause()
+                        } else {
+                            player.play()
+                        }
+                    } else {
+                        if (voiceCache.isInCache()) {
+                            player.acquireControlAndPlay()
+                        } else {
+                            scope.launch { downloadCacheAndPlay() }
+                        }
+                    }
```
The ownership aspect is a lot for the presenter to know about. Could the `VoiceMessagePlayer` not take care of some of this logic? Leaving us with something simpler like

```kotlin
if (!isDownloaded) {
   download()
}

if (isPlaying) {
   pause()
} else {
   play()
}
```

julioromano replied:
I agree, but I prefer to take some more time aside to make this better.

jonnyandrew replied:
Of course, I don't think it should block this PR

jonnyandrew replied:
Of course, I don't think it should block this PR

jonnyandrew commented on features/messages/impl/src/main/kotlin/io/element/android/features/messages/impl/voicemessages/timeline/VoiceMessagePresenter.kt:
```diff
@@ -0,0 +1,152 @@
+        }
+        val time by remember {
+            derivedStateOf {
+                val time = if (playerState.isMyMedia) playerState.currentPosition else content.duration.toMillis()
+                "%02d:%02d".format(time.msToMins(), time.msToSecs())
```
FYI we have a utility for this formatting: `Duration.formatShort()`.

julioromano replied:
Yeah now we do :) switched to it.

jonnyandrew commented on features/messages/impl/src/main/kotlin/io/element/android/features/messages/impl/voicemessages/timeline/WaveformProgressIndicator.kt:
```diff
@@ -0,0 +1,84 @@
+        onProgressChangeFinished = {
+            // This is to send just one onSeek callback after the user has finished seeking.
+            // Otherwise the AudioWaveform library would send multiple callbacks while the user is seeking.
+            val p = seekProgress!!
```
Instead of crashing with NPE can we log an error and fail silently?

If it's an important error, we could also track it to Sentry?

julioromano replied:
It is my understanding that `onProgressChangeFinished` will always be called after at least one invocation of `onProgressChange` so NPEs should be impossible here.

jonnyandrew replied:
If it certainly is the case, no problem. Personally I would er on the side of caution!

jonnyandrew replied:
If it certainly is the case, no problem. Personally I would er on the side of caution!

jonnyandrew commented on features/messages/impl/src/main/kotlin/io/element/android/features/messages/impl/voicemessages/timeline/WaveformProgressIndicator.kt:
```diff
@@ -0,0 +1,84 @@
+internal fun WaveformProgressIndicatorPreview() = ElementPreview {
+    WaveformProgressIndicator(
+        progress = 0.5f,
+        amplitudes = persistentListOf(0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 8, 7, 6, 5, 4, 3, 2, 1, 0),
```
nit: Perhaps some more examples with different amplitude list lengths?

jonnyandrew commented on tests/uitests/src/test/snapshots/images/ui_S_t[f.messages.impl.timeline.components.event_null_TimelineItemVoiceView-D-39_39_null_1,NEXUS_5,1.0,en].png:
nit: Is there any way these previews could be put into a single screenshot? It'd make it easier to review and compare spacing between the different states.

julioromano replied:
Done in `TimelineItemVoiceViewUnifiedPreview` though I don't see the waveform here. I suspect the library that draws it does not show it on first pass so it doesn't appear in the screenshot.

jonnyandrew reviewed (commented):
Looks good! I haven't gotten to review the tests yet. Here's my comments in the mean time

julioromano commented on features/messages/impl/src/main/kotlin/io/element/android/features/messages/impl/timeline/model/event/TimelineItemVoiceContentProvider.kt:
```diff
@@ -0,0 +1,49 @@
+        get() = sequenceOf(
+            aTimelineItemVoiceContent(body = "A sound.mp3"),
+            aTimelineItemVoiceContent(body = "A bigger name sound.mp3"),
+            aTimelineItemVoiceContent(body = "An even bigger bigger bigger bigger bigger bigger bigger sound name which doesn't fit .mp3"),
```
This provider is actually unused, I initially created it to match `TimelineItemAudioContentProvider` but then when adding `VoiceMessagePresenter` I opted to use `VoiceMessageStateProvider` in `TimelineItemVoiceView` instead.
Thing is a preview can only use one provider (instead here I'd love to have 2) so I kind of forgot about it over time.
I know created `TimelineItemVoiceViewParameterProvider` which combines the 2 so I can use that in the preview instead.

But I still think there is still work to do as refinement here, for instance: should we keep the separate `content` object or should we just embed it in the state and only expose the state on the view side? @ganfra I'd like to also have your take on this.

julioromano commented on features/messages/impl/src/main/kotlin/io/element/android/features/messages/impl/voicemessages/timeline/VoiceMessageCache.kt:
```diff
@@ -0,0 +1,123 @@
+     *
+     * @return the file path of the voice message in the cache directory.
+     */
+    val cachePath: String
```
That's a file path (as in full path to file e.g. `/a/b/c/d/something.ogg`) not a directory.
I get it was not clear enough from the docstring, is there a better way to explain it in English?

jonnyandrew replied:
I see - I was a bit confused when I first saw the interface what it's for. I thought this is used to provide access to the file. But I think it's used to create the mediaUri [here](https://github.com/vector-im/element-x-android/pull/1503/files/b23f1d48f553ecaf51c0c6a37a14e6b842eb3aab#diff-ead1cbb0afc8a632d8782728759c8db7c31fd5abbc3412d7407c127aac4e7d06R132)?

It's not a big problem if we need it in this format - happy to leave as is!

julioromano replied:
Both.
The idea is that we map one mxc:// content uri to one file in the cache via a simple mapping: `"${cacheDir.path}/$CACHE_VOICE_SUBDIR/${mxcUri2FilePath(mxcUri)}"` .

So once we have a content uri we can initialize an instance of this class and:
1) it will give us the path to the cache file using `cachePath`
2) it will tell us if the file at such path actually exists with `isInCache()` (i.e. cache hit) or not (i.e. cache miss)
3) it will give use the option to fill the cache with such content (`moveToCache()`)

Perhaps we can find better method names?

julioromano commented on features/messages/impl/src/main/kotlin/io/element/android/features/messages/impl/voicemessages/timeline/VoiceMessageCache.kt:
```diff
@@ -0,0 +1,123 @@
+ *
+ * See: https://spec.matrix.org/v1.8/client-server-api/#matrix-content-mxc-uris
+ */
+private val mxcRegex = Regex("""^mxc:\/\/([^\/]+)\/([^\/]+)$""")
```
Not sure, that module is only to interface with the rust sdk AFAIK. @bmarty any suggestions?

jonnyandrew commented on features/messages/impl/src/main/kotlin/io/element/android/features/messages/impl/timeline/components/event/TimelineItemVoiceView.kt:
```diff
@@ -0,0 +1,206 @@
+) {
+    IconButton(
+        drawableRes = R.drawable.play,
+        contentDescription = null, // TODO
```
We discussed on the designs about the [alt text for the voice recording UI](https://www.figma.com/file/0MMNu7cTOzLOlWb7ctTkv3?node-id=11987:104884&mode=dev#582517193) so I assume the same applies for the timeline. Perhaps one to check with @callumu though if you're not sure.

I can see `a11y.play` and a11y.pause` strings are there in Localazy already. But if translations don't exist feel free to create them (and good to make sure we reuse the same across Android and iOS).

julioromano replied:
Done, there was already an `action_retry` string so I used that one for the retry button.

julioromano commented:
> Looks good! I haven't gotten to review the tests yet. Here's my comments in the mean time

Pushed first batch of fixes, there are still a few to go.

jonnyandrew commented on features/messages/impl/src/test/kotlin/io/element/android/features/messages/mediaplayer/MediaPlayerFake.kt:
```diff
@@ -0,0 +1,71 @@
+/**
+ * Fake implementation of [MediaPlayer] for testing purposes.
+ */
+class MediaPlayerFake : MediaPlayer {
```
I think the naming convention we're using is `FakeClassName` e.g. `FakeMediaPlayer`

julioromano replied:
Changed!

jonnyandrew commented on features/messages/impl/src/test/kotlin/io/element/android/features/messages/voicemessages/play/VoiceMessageCacheFake.kt:
Should the package be updated here from `play` to `timeline`?

julioromano replied:
Oh yesss!

julioromano commented:
Addressed all suggestions except: https://github.com/vector-im/element-x-android/pull/1503#discussion_r1370466839
For which I'd like to take some more time to come up with a better solution.

jonnyandrew reviewed (approved):
Thanks for the changes. I don't think we have any blocking issues here. Looking forward to merging this!
